sea-orm-macros/src/derives/value_type.rs

```diff
@@ -0,0 +1,240 @@
+
+struct DeriveValueType {
+    name: syn::Ident,
+    ty: Type,
+    column_type: TokenStream,
+    array_type: TokenStream,
+}
+
+impl DeriveValueType {
+    pub fn new(input: syn::DeriveInput) -> Option<Self> {
+        let dat = input.data;
+        let fields: Option<syn::punctuated::Punctuated<syn::Field, syn::token::Comma>> = match dat {
+            syn::Data::Struct(syn::DataStruct {
+                fields: syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }),
+                ..
+            }) => Some(unnamed),
+            _ => None,
+        };
+
+        let field = fields
+            .expect("This derive accept only struct")
+            .first()
+            .expect("The struct should contain one value field")
+            .to_owned();
+
+        let name = input.ident;
+        let mut col_type = None;
+        let mut arr_type = None;
+
+        for attr in input.attrs.iter() {
+            if !attr.path().is_ident("sea_orm") {
+                continue;
+            }
+
+            attr.parse_nested_meta(|meta| {
+                if meta.path.is_ident("column_type") {
+                    let lit = meta.value()?.parse()?;
+                    if let Lit::Str(litstr) = lit {
+                        let ty: TokenStream = syn::parse_str(&litstr.value())?;
+                        col_type = Some(ty);
+                    } else {
+                        return Err(meta.error(format!("Invalid column_type {:?}", lit)));
+                    }
+                } else if meta.path.is_ident("array_type") {
+                    let lit = meta.value()?.parse()?;
+                    if let Lit::Str(litstr) = lit {
+                        let ty: TokenStream = syn::parse_str(&litstr.value())?;
+                        arr_type = Some(ty);
+                    } else {
+                        return Err(meta.error(format!("Invalid array_type {:?}", lit)));
+                    }
+                } else {
+                    // received other attribute
+                    return Err(meta.error(format!("Invalid attribute {:?}", meta.path)));
+                }
+
+                Ok(())
+            })
+            .unwrap_or(());
+        }
+
+        let ty = field.clone().ty;
+        let field_type = quote! { #ty }
+            .to_string() //E.g.: "Option < String >"
+            .replace(' ', ""); // Remove spaces
+        let field_type = if field_type.starts_with("Option<") {
+            &field_type[7..(field_type.len() - 1)] // Extract `T` out of `Option<T>`
+        } else {
+            field_type.as_str()
+        };
+
+        let column_type = match col_type {
+            Some(t) => quote! { sea_orm::sea_query::ColumnType::#t },
+            None => {
+                let col_type = match field_type {
+                    "char" => quote! { Char(None) },
+                    "String" | "&str" => quote! { String(None) },
+                    "i8" => quote! { TinyInteger },
```

Comment:
Umm... we should not copy and paste here. Can we extract the function somewhere and share the same code among different macros?